RFC Draft.md

```diff
@@ -377,9 +377,339 @@ const uploadedObjectReference = await uploadTask.result;
+import { CreateTodoInput } from './API';
+// so many imports from disparate modules 
+
+function createMutation() {
+  const createInput: CreateTodoInput = {
+    name: 'Improve API TS support ',
+  };
+
+  const res = await API.mutate(createTodo, {
+    input: createInput,
+  });
+
+  // The returned data is the result of the request. If there are more than one queries/mutations in a request,
+  // then the return value stays the same as v5. i.e. res.createTodo.data
+  const newTodo = res;
+}
+
+function subscribeToCreate() {
+  const sub = API.subscribe(onCreateTodo).on({
+    next: (message) => {
+      // Return value immediately available instead of `value?.data?.onCreateTodo`
+      const newTodo = message;
+    }
+  });
+}
+```
+
+## Flatten GraphQL operation responses
+As alluded to in the previous section, we're looking to flatten the results of GraphQL operations to make them more easily accessible instead of the current three-levels-deep nested object. Would love to get your understanding on which option you prefer.
+
+**Current Usage (v5)**
+```ts
+async function createNewTodo() {
+  const res: GraphQLResult<Todo> = await API.graphql<GraphQLQuery<Todo>>(
+    graphqlOperation(createTodo, {
+      input: { id: uuid() },
+    })
+  );
+
+  // Mutation result is nested
+  console.log(res.data.createTodo); 
+}
+```
+
+### Proposed behavior for single query/mutation in the response
+**Proposed Option 1: Flatten to the lowest level (v6)**
+```ts
+async function createNewTodo() {
+  const res: Todo = await API.mutate(createTodo, {
+    input: { id: uuid() },
+  });
+  
+  // Response flattened to the todo level
+  console.log(res); 
+}
+```
+
+**Proposed Option 2: Flatten to the `data` level (v6)**
+```ts
+async function createNewTodo() {
+  const res: Todo = await API.mutate(createTodo, {
+    input: { id: uuid() },
+  });
+
+  // Response flattened to the `data` level
+  console.log(res.createTodo); 
+}
+
+interface GraphQLData<T = object> {
+  [query: string]: T // in the above example T is Todo
+}
+```
+
+### Proposed behavior for multiple queries/mutations in the response
+In GraphQL, you can define multiple queries or mutations in a single request. The response object will include the result of all the queries and mutations. For example, given the following queries:
+
+```ts
+async function custom() {
+  const operation = {
+    query: `
+      query GetTodo($todoId: ID!, $fooId: ID!) {
+        getTodo(id: $todoId) {
+          id
+          name
+          createdAt
+          updatedAt
+        }
+        getFoo(id: $fooId) {
+          id
+          name
+          createdAt
+          updatedAt
+        }
+      }
+    `,
+    variables: { todoId: 'c48481bd-f808-426f-8fed-19e1368ca0bc', fooId: '9d4e6e30-fcb4-4409-8160-7d44931a6a02' },
+    authToken: undefined,
+    userAgentSuffix: undefined
+  }
+  
+  const result = await API.query<MyQueryType>(operation.query, {
+    variables: operation.variables
+  });
```

Comment:
I must have missed this previously. But, seeing this written out this way, I feel this is not compatible with the fine-grained graphql above. I.e., until this point in the RFC, we've still been operating under the assumption that `query` is a single query/mutation.

Reply:
Actually, maybe I misread this a little. This might be fine.

Reply:
```suggestion
  const result = await API.graphql(operation.query, {
    variables: operation.variables
  });
```

You're right. I believe when we chatted on the call, we said that we'd retain the `.graphql` for this use case.